src/org/jetbrains/plugins/scala/lang/refactoring/util/ScalaTypeValidator.scala

```diff
@@ -64,43 +53,34 @@ class ScalaTypeValidator(conflictsReporter: ConflictsReporter,
 
     buf ++= getForbiddenNames(container, name)
 
-    //    val parent = container.getContext
-    buf ++= getForbiddenNamesInBlock(container, name)
+    if (buf.isEmpty) {
+      buf ++= getForbiddenNamesInBlock(container, name)
+    }
+
     buf.toArray
   }
 
   //TODO maybe not the best way to handle with such matching
```

Comment:
You still have recursion here: matchElement calls getForbiddenNamesInBlock and vice versa. Second method can be removed completely.